<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Fabio Morgante &amp; Deborah Kunkel" />

<meta name="date" content="2024-02-13" />

<title>Analysis of blood cell traits in UK Biobank</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">mr_mash_rss_analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/morgante_lab/mr_mash_rss_analysis">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Analysis of blood cell traits in UK
Biobank</h1>
<h4 class="author">Fabio Morgante &amp; Deborah Kunkel</h4>
<h4 class="date">February 13, 2024</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-02-13
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>mr_mash_rss/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20230612code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20230612)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20230612code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20230612)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcommorgantelabmrmashrssanalysistree9383eec5b7094e531fe07800531ea3103b8e4e44targetblank9383eeca">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/tree/9383eec5b7094e531fe07800531ea3103b8e4e44" target="_blank">9383eec</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcommorgantelabmrmashrssanalysistree9383eec5b7094e531fe07800531ea3103b8e4e44targetblank9383eeca"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/tree/9383eec5b7094e531fe07800531ea3103b8e4e44" target="_blank">9383eec</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .snakemake/
    Ignored:    data/
    Ignored:    output/
    Ignored:    run/
    Ignored:    tmp/

Untracked files:
    Untracked:  Snakefile_old
    Untracked:  code/adjust_LD_test.R
    Untracked:  code/compute_sumstats_test_bc_ukb.R
    Untracked:  code/fit_kriging_rss_ukb.R
    Untracked:  code/fit_mrmashrss_test_ukb.R
    Untracked:  code/merge_sumstats_all_chr_ukb.R
    Untracked:  code/split_effects_by_chr_imp_ukb.R
    Untracked:  scripts/11_run_fit_mrmashrss_by_chr_V_all_chr_bc_ukb.sbatch
    Untracked:  scripts/11_run_fit_mrmashrss_sparse_LD_mvsusie_paper_prior_by_chr_bc_ukb.sbatch
    Untracked:  scripts/12_run_compute_pred_accuracy_mrmashrss_sparse_LD_mvsusie_paper_prior_bc_ukb.sbatch
    Untracked:  scripts/12_run_fit_mrmashrss_sparse_LD_Vcor_all_chr_init_by_chr_bc_ukb.sbatch
    Untracked:  scripts/13_run_compute_pred_accuracy_mrmashrss_sparse_LD_Vcor_all_chr_init_bc_ukb.sbatch
    Untracked:  scripts/3_run_compute_sumstats_by_chr_and_fold_test_bc_ukb.sbatch
    Untracked:  scripts/7_run_compute_residual_cor_all_chr_bc_ukb.sbatch
    Untracked:  scripts/run_adjust_LD_test.sbatch
    Untracked:  scripts/run_fit_kriging_rss_by_chr_and_trait_bc_ukb.sbatch

Unstaged changes:
    Modified:   scripts/12_75_run_fit_mrmashrss_sparse_LD_V_all_chr_init_prior_finemapped_by_chr_bc_sampled_ukb.sbatch

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/ukb_bc_results.Rmd</code>) and
HTML (<code>docs/ukb_bc_results.html</code>) files. If you’ve configured
a remote Git repository (see <code>?wflow_git_remote</code>), click on
the hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/9383eec5b7094e531fe07800531ea3103b8e4e44/analysis/ukb_bc_results.Rmd" target="_blank">9383eec</a>
</td>
<td>
fmorgante
</td>
<td>
2024-02-13
</td>
<td>
Add BayesR results with small sample size
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/adb9a12b1fa3c4419cf04ff5536785aad04c615c/analysis/ukb_bc_results.Rmd" target="_blank">adb9a12</a>
</td>
<td>
fmorgante
</td>
<td>
2024-02-09
</td>
<td>
Allow for mr.mash.rss to be initialized using any other method
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/60889dd871514919b74d43907d52d9ee47e06821/docs/ukb_bc_results.html" target="_blank">60889dd</a>
</td>
<td>
fmorgante
</td>
<td>
2024-02-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/36dc3cfd44741d8dc17962ba00b2850bf7065669/analysis/ukb_bc_results.Rmd" target="_blank">36dc3cf</a>
</td>
<td>
fmorgante
</td>
<td>
2024-02-04
</td>
<td>
Add results with smaller sample size
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/6e40b7a039c4a85d84cf03cb8f1af2582a9f3b0c/docs/ukb_bc_results.html" target="_blank">6e40b7a</a>
</td>
<td>
fmorgante
</td>
<td>
2024-01-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/87d67925ed1d07199ec79b00f79de35b51f33bc9/analysis/ukb_bc_results.Rmd" target="_blank">87d6792</a>
</td>
<td>
fmorgante
</td>
<td>
2024-01-02
</td>
<td>
Add results with finemapped prior
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/11d9b6c6c4d17681e62e388fa571c65c68e69b73/docs/ukb_bc_results.html" target="_blank">11d9b6c</a>
</td>
<td>
fmorgante
</td>
<td>
2023-11-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/e2a7fe46e0d6f066b5ffaf30e4e3a6b9b0d636fb/analysis/ukb_bc_results.Rmd" target="_blank">e2a7fe4</a>
</td>
<td>
fmorgante
</td>
<td>
2023-11-13
</td>
<td>
Add description of the results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/8417b2fbf36ee2cac2dd160119c1fc4f925d4b2e/docs/ukb_bc_results.html" target="_blank">8417b2f</a>
</td>
<td>
fmorgante
</td>
<td>
2023-11-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/c254c89fee422037c493937c7cd39cdf47e1f918/analysis/ukb_bc_results.Rmd" target="_blank">c254c89</a>
</td>
<td>
fmorgante
</td>
<td>
2023-11-13
</td>
<td>
Try fixing a bug
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/a0f2ec934d229526f49e6a3c88e9b62c3c710b27/docs/ukb_bc_results.html" target="_blank">a0f2ec9</a>
</td>
<td>
fmorgante
</td>
<td>
2023-11-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/5bd74b48cb57a82c08d32015696a1333277dcd0b/analysis/ukb_bc_results.Rmd" target="_blank">5bd74b4</a>
</td>
<td>
fmorgante
</td>
<td>
2023-11-13
</td>
<td>
Add description of new methods
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/090c32310a8fedf6247c37fbcd6d32c6deac4c49/docs/ukb_bc_results.html" target="_blank">090c323</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-30
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/7822773394720f8ba7cb343f993b17bf6d4f602e/analysis/ukb_bc_results.Rmd" target="_blank">7822773</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-30
</td>
<td>
Adjust labels
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/9ad448b9e9b4ceaf44d89f9ec737cef07861db93/docs/ukb_bc_results.html" target="_blank">9ad448b</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-30
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/89d9d1fb56b77329fe53eea4624d9f0a19dc1b4d/analysis/ukb_bc_results.Rmd" target="_blank">89d9d1f</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-30
</td>
<td>
Add name of the traits
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/2ffac98d0e1de07b2c45bcd2b5acaa517b18db1c/docs/ukb_bc_results.html" target="_blank">2ffac98</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-30
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/af4c5e8abd652e68865ec05fa52066226ad19696/analysis/ukb_bc_results.Rmd" target="_blank">af4c5e8</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-30
</td>
<td>
Remove ldsc from plots
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/8520d64f5e6ad8046c9be746bba9f9b61ceba68b/docs/ukb_bc_results.html" target="_blank">8520d64</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-30
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/feefb657d4f1f846a47a3c76a7120914001cedc2/analysis/ukb_bc_results.Rmd" target="_blank">feefb65</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-30
</td>
<td>
Add mr.mash.rss results
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/7da29c7ad49795d31e3d9e1fe0228c5541c98cb9/docs/ukb_bc_results.html" target="_blank">7da29c7</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/febf669ca5f7d522297abf9c8dec9f75f7ee300a/analysis/ukb_bc_results.Rmd" target="_blank">febf669</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-26
</td>
<td>
Fix another bug
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/morgantelab/mr_mash_rss_analysis/7cd0ded97c2d6b525cb49fbf68f79009168e4290/docs/ukb_bc_results.html" target="_blank">7cd0ded</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/4b1fbf324d67c13e02932a8406230fa62d470eb5/analysis/ukb_bc_results.Rmd" target="_blank">4b1fbf3</a>
</td>
<td>
fmorgante
</td>
<td>
2023-10-26
</td>
<td>
Add initial bc results
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>###Load libraries
library(ggplot2)
library(cowplot)

prefix &lt;- &quot;output/prediction_accuracy/ukb_bc&quot;
prefix_2 &lt;- &quot;output/prediction_accuracy/ukb_bc_sampled&quot;</code></pre>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The goal of this analysis is to benchmark the newly developed
<em>mr.mash.rss</em> (aka <em>mr.mash</em> with summary data) against
already existing methods in the task of predicting phenotypes from
genotypes using only summary data. After performing several simulations,
we analyzed analysis 16 blood cell traits in UK Biobank.</p>
</div>
<div id="description-of-the-analysis" class="section level2">
<h2>Description of the analysis</h2>
<p>Following the <a
href="https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1">mvSuSiE
paper</a>, we selected 16 blood cell traits from the total available in
the UK Biobank haematology data collection and after the filtering
steps, we were left with 244,049 samples.</p>
<p>A 5-fold cross validation scheme was employed, whereby the samples
were randomly divided in 5 disjoint sets. At each iteration, 4 sets were
used as training set and 1 set was used as test set.</p>
<p>Summary statistics were computed in the training set by first
regressing out the effect of sex, age at recruitment, age <span
class="math inline">\(\times\)</span> age, assessment centre, and
genotype measurement batch, and the top 10 genotype PCs using a linear
model. Then, we ran a GWAS using a simple linear regression on the
quantile normalized residuals from the previous step.</p>
<p>The sparse LD matrix was computed in the training set as described in
the <a
href="https://academic.oup.com/bioinformatics/article/36/22-23/5424/6039173">LDpred2
paper</a></p>
<p>For the majority of the analyses, we used a use a set of 1,054,330
HapMap3 variants. Only for the computation of the data-driven covariance
matrices for <em>mr.mash.rss</em> we used a larger sets of variants as
described in the <a
href="https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1">mvSuSiE
paper</a>.</p>
<p>Two different methods were fitted to the summary statistics:</p>
<ul>
<li><p><a
href="https://academic.oup.com/bioinformatics/article/36/22-23/5424/6039173">LDpred2</a>
per-chromosome with the auto option, 1000 iterations (after 500 burn-in
iterations), <span class="math inline">\(h^2\)</span> initialized using
an estimate from LD Score regression (LDSC) and <span
class="math inline">\(p\)</span> initialized using the same grid as in
the original paper. NB this is a univariate method.</p></li>
<li><p><em>mr.mash.rss</em> per-chromosome, with both canonical and
data-driven covariance matrices computed as described in the <a
href="https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1">mvSuSiE
paper</a>, updating the (full rank) residual covariance and the mixture
weights, without standardizing the variables. The residual covariance
was initialized as in the <a
href="https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1">mvSuSiE
paper</a> and the mixture weights were initialized as 90% of the weight
on the null component and 10% of the weight split equally across the
remaining components. The phenotypic covariance was computed as the
sample covariance using the individual-level data. NB this is a
multivariate method.</p></li>
<li><p><em>mr.mash.rss</em> per-chromosome, as above, but posterior mean
of regression coeffcients initialized using the estimates from
<em>LDpred2</em>. NB this is a multivariate method.</p></li>
<li><p><em>mr.mash.rss</em> per-chromosome, as above, but posterior mean
of regression coeffcients initialized using the estimates from
<em>LDpred2</em>, and using finemapped variants (via SuSiE-RSS) to
compute the data-driven covariance matrices. NB this is a multivariate
method.</p></li>
</ul>
<p>Prediction accuracy was evaluated as the <span
class="math inline">\(R^2\)</span> of the regression of (quantile
normalized) true phenotypes on the predicted phenotypes in the test set.
This metric as the attractive property that its upper bound is <span
class="math inline">\(h_g^2\)</span>. Here we report the results for
each fold.</p>
</div>
<div id="results-with-244k-individuals" class="section level2">
<h2>Results with 244K individuals</h2>
<pre class="r"><code>dat_ldsc &lt;- matrix(as.numeric(NA), 5, 16)

for(i in 1:5){
  for(s in 1:16){
    dat_ldsc[i, s] &lt;- readRDS(paste0(&quot;output/ldsc_fit/ukb_bc_chrAll_ldsc_fit_trait&quot;, s, &quot;_&quot;, i, &quot;.rds&quot;))[&quot;h2&quot;]
  }
  
  if(i&gt;1){
    dat_mrmash &lt;- rbind(dat_mrmash, readRDS(paste0(prefix, &quot;_mr_mash_rss_sparse_LD_V_all_chr_pred_acc_&quot;, i, &quot;.rds&quot;))$r2)
    dat_mrmash_init &lt;- rbind(dat_mrmash_init, readRDS(paste0(prefix, &quot;_mr_mash_rss_sparse_LD_V_all_chr_init_pred_acc_&quot;, i, &quot;.rds&quot;))$r2)
    dat_mrmash_init_fm_prior &lt;- rbind(dat_mrmash_init_fm_prior, readRDS(paste0(prefix, &quot;_mr_mash_rss_sparse_LD_V_all_chr_init_prior_finemapped_pred_acc_&quot;, i, &quot;.rds&quot;))$r2)
    dat_ldpred2 &lt;- rbind(dat_ldpred2, readRDS(paste0(prefix, &quot;_ldpred2_auto_pred_acc_&quot;, i, &quot;.rds&quot;))$r2)
    dat_bayesR &lt;- rbind(dat_bayesR, readRDS(paste0(prefix, &quot;_bayesR_pred_acc_&quot;, i, &quot;.rds&quot;))$r2)
  } else {
    dat_mrmash &lt;- readRDS(paste0(prefix, &quot;_mr_mash_rss_sparse_LD_V_all_chr_pred_acc_&quot;, i, &quot;.rds&quot;))$r2
    dat_mrmash_init &lt;- readRDS(paste0(prefix, &quot;_mr_mash_rss_sparse_LD_V_all_chr_init_pred_acc_&quot;, i, &quot;.rds&quot;))$r2
    dat_mrmash_init_fm_prior &lt;- readRDS(paste0(prefix, &quot;_mr_mash_rss_sparse_LD_V_all_chr_init_prior_finemapped_pred_acc_&quot;, i, &quot;.rds&quot;))$r2
    dat_ldpred2 &lt;- readRDS(paste0(prefix, &quot;_ldpred2_auto_pred_acc_&quot;, i, &quot;.rds&quot;))$r2
    dat_bayesR &lt;- readRDS(paste0(prefix, &quot;_bayesR_pred_acc_&quot;, i, &quot;.rds&quot;))$r2
  }
}

means_mrmash &lt;- colMeans(dat_mrmash)
means_mrmash_init &lt;- colMeans(dat_mrmash_init)
means_mrmash_init_fm_prior &lt;- colMeans(dat_mrmash_init_fm_prior)
means_ldpred2 &lt;- colMeans(dat_ldpred2)
means_bayesR &lt;- colMeans(dat_bayesR)

perc_change &lt;- ((means_mrmash_init-means_ldpred2)/means_ldpred2)*100
perc_change_fm_prior &lt;- ((means_mrmash_init_fm_prior-means_ldpred2)/means_ldpred2)*100
perc_change_bayesR_fm_prior &lt;- ((means_mrmash_init_fm_prior-means_bayesR)/means_bayesR)*100

pheno &lt;- readRDS(&quot;data/phenotypes/ukb_cleaned_bc_adjusted_pheno_test_1.rds&quot;)

names(perc_change) &lt;- colnames(pheno)
names(perc_change_fm_prior) &lt;- colnames(pheno)
names(perc_change_bayesR_fm_prior) &lt;- colnames(pheno)

linez &lt;- data.frame(trait=colnames(pheno),
                    h2=colMeans(dat_ldsc))

r2 &lt;- c(as.vector(dat_mrmash), as.vector(dat_mrmash_init), as.vector(dat_mrmash_init_fm_prior), 
        as.vector(dat_ldpred2), as.vector(dat_bayesR))
method &lt;- rep(c(&quot;mr_mash_rss&quot;, &quot;mr_mash_rss_init&quot;, &quot;mr_mash_rss_init_fm_prior&quot;, &quot;ldpred2_auto&quot;, &quot;bayesR&quot;), each=80)
trait &lt;- rep(colnames(pheno), each=5)

res &lt;- data.frame(method, trait, r2)

res &lt;- res[which(res$method %in% c(&quot;bayesR&quot;, &quot;ldpred2_auto&quot;, &quot;mr_mash_rss&quot;, &quot;mr_mash_rss_init&quot;, &quot;mr_mash_rss_init_fm_prior&quot;)), ]

res &lt;- transform(res,
                 method=factor(method, levels=c(&quot;bayesR&quot;, &quot;ldpred2_auto&quot;, 
                                                &quot;mr_mash_rss&quot;, &quot;mr_mash_rss_init&quot;, &quot;mr_mash_rss_init_fm_prior&quot;),
                               labels=c(&quot;BayesR&quot;, &quot;LDpred2&quot;, &quot;mr.mash.rss&quot;, &quot;mr.mash.rss init&quot;, &quot;mr.mash.rss init prior&quot;)),
                 trait=factor(trait))

p_methods_shared &lt;- ggplot(res, aes(x = trait, y = r2, fill = method)) +
  geom_boxplot(color = &quot;black&quot;, outlier.size = 1, width = 0.85) +
  stat_summary(fun=mean, geom=&quot;point&quot;, shape=23,
               position = position_dodge2(width = 0.87,   
                                          preserve = &quot;single&quot;)) +
  scale_fill_manual(values = c(&quot;pink&quot;, &quot;red&quot;, &quot;green&quot;, &quot;orange&quot;, &quot;blue&quot;)) +
  labs(x = &quot;Trait&quot;, y = expression(italic(R)^2), fill=&quot;Method&quot;, title=&quot;&quot;) + 
  # facet_grid(~trait, scales=&quot;free_x&quot;) +
  # geom_hline(aes(yintercept = h2), linez) +
  theme_cowplot(font_size = 18) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

print(p_methods_shared)</code></pre>
<p><img src="figure/ukb_bc_results.Rmd/results_244K_accuracy-1.png" width="1248" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results_244K_accuracy-1">
Past versions of results_244K_accuracy-1.png
</button>
</p>
<div id="fig-results_244K_accuracy-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/60889dd871514919b74d43907d52d9ee47e06821/docs/figure/ukb_bc_results.Rmd/results_244K_accuracy-1.png" target="_blank">60889dd</a>
</td>
<td>
fmorgante
</td>
<td>
2024-02-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
<!-- The percentage change in $R^2$ of *mr.mash.rss init* over *LDpred2* by trait is:  -->
<!-- ```{r perc_change} -->
<!-- perc_change -->
<!-- ``` -->
<p>The percentage change in <span class="math inline">\(R^2\)</span> of
<em>mr.mash.rss init prior</em> over <em>BayesR</em> by trait is:</p>
<pre class="r"><code>perc_change_bayesR_fm_prior</code></pre>
<pre><code>        WBC_count         RBC_count       Haemoglobin               MCV 
        1.0764560         1.0705411        -1.3652685         1.3731148 
              RDW    Platelet_count      Plateletcrit               PDW 
        3.1868160         1.1206196         0.5615176         0.6856654 
  Lymphocyte_perc     Monocyte_perc  Neutrophill_perc  Eosinophill_perc 
        5.2227381         1.0893654         7.4217879        -2.2419269 
   Basophill_perc Reticulocyte_perc              MSCV          HLR_perc 
       13.0740144         3.4547098         4.6498409         1.4944277 </code></pre>
<!-- The overall percentage change in $R^2$ of *mr.mash.rss init* over *LDpred2* across traits is 5.1217612. -->
<p>The overall percentage change in <span
class="math inline">\(R^2\)</span> of <em>mr.mash.rss init prior</em>
over <em>BayesR</em> across traits is 2.6171512.</p>
<!-- The results show that *mr.mash.rss init* improves accuracy over *LDpred2* for all but one trait. However, -->
<!-- in that trait the performance is very similar. In general, the improvement in performance is in line with -->
<!-- other papers such as the [PRS-CS paper](https://www.nature.com/articles/s41467-019-09718-5). -->
</div>
<div id="results-with-75k-individuals" class="section level2">
<h2>Results with 75K individuals</h2>
<p>We hypothesized that the advantage of a multivariate analysis would
be more pronounced with a smaller sample size. Thus, we randomly sampled
75,000 individuals from the 244,000 and repeated the analyses above.</p>
<pre class="r"><code>for(i in 1:5){
  if(i&gt;1){
    dat_mrmash_init_fm_prior &lt;- rbind(dat_mrmash_init_fm_prior, readRDS(paste0(prefix_2, &quot;_mr_mash_rss_sparse_LD_V_all_chr_init_prior_finemapped_pred_acc_&quot;, i, &quot;.rds&quot;))$r2)
    dat_ldpred2 &lt;- rbind(dat_ldpred2, readRDS(paste0(prefix_2, &quot;_ldpred2_auto_pred_acc_&quot;, i, &quot;.rds&quot;))$r2)
    dat_bayesR &lt;- rbind(dat_bayesR, readRDS(paste0(prefix_2, &quot;_bayesR_pred_acc_&quot;, i, &quot;.rds&quot;))$r2)
  } else {
    dat_mrmash_init_fm_prior &lt;- readRDS(paste0(prefix_2, &quot;_mr_mash_rss_sparse_LD_V_all_chr_init_prior_finemapped_pred_acc_&quot;, i, &quot;.rds&quot;))$r2
    dat_ldpred2 &lt;- readRDS(paste0(prefix_2, &quot;_ldpred2_auto_pred_acc_&quot;, i, &quot;.rds&quot;))$r2
    dat_bayesR &lt;- readRDS(paste0(prefix_2, &quot;_bayesR_pred_acc_&quot;, i, &quot;.rds&quot;))$r2
  }
}

means_mrmash_init_fm_prior &lt;- colMeans(dat_mrmash_init_fm_prior)
means_ldpred2 &lt;- colMeans(dat_ldpred2)
means_bayesR &lt;- colMeans(dat_bayesR)

perc_change_fm_prior &lt;- ((means_mrmash_init_fm_prior-means_ldpred2)/means_ldpred2)*100
perc_change_bayesR_fm_prior &lt;- ((means_mrmash_init_fm_prior-means_bayesR)/means_bayesR)*100

pheno &lt;- readRDS(&quot;data/phenotypes/ukb_cleaned_bc_adjusted_pheno_test_1.rds&quot;)

names(perc_change_fm_prior) &lt;- colnames(pheno)
names(perc_change_bayesR_fm_prior) &lt;- colnames(pheno)

r2 &lt;- c(as.vector(dat_mrmash_init_fm_prior), as.vector(dat_ldpred2), as.vector(dat_bayesR))
method &lt;- rep(c(&quot;mr_mash_rss_init_fm_prior&quot;, &quot;ldpred2_auto&quot;, &quot;bayesR&quot;), each=80)
trait &lt;- rep(colnames(pheno), each=5)

res &lt;- data.frame(method, trait, r2)

res &lt;- res[which(res$method %in% c(&quot;bayesR&quot;, &quot;ldpred2_auto&quot;, &quot;mr_mash_rss_init_fm_prior&quot;)), ]

res &lt;- transform(res,
                 method=factor(method, levels=c(&quot;bayesR&quot;, &quot;ldpred2_auto&quot;, &quot;mr_mash_rss_init_fm_prior&quot;),
                               labels=c(&quot;BayesR&quot;, &quot;LDpred2&quot;, &quot;mr.mash.rss init prior&quot;)),
                 trait=factor(trait))

p_methods_shared &lt;- ggplot(res, aes(x = trait, y = r2, fill = method)) +
  geom_boxplot(color = &quot;black&quot;, outlier.size = 1, width = 0.85) +
  stat_summary(fun=mean, geom=&quot;point&quot;, shape=23,
               position = position_dodge2(width = 0.87,   
                                          preserve = &quot;single&quot;)) +
  scale_fill_manual(values = c(&quot;pink&quot;, &quot;red&quot;, &quot;blue&quot;)) +
  labs(x = &quot;Trait&quot;, y = expression(italic(R)^2), fill=&quot;Method&quot;, title=&quot;&quot;) + 
  # facet_grid(~trait, scales=&quot;free_x&quot;) +
  # geom_hline(aes(yintercept = h2), linez) +
  theme_cowplot(font_size = 18) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

print(p_methods_shared)</code></pre>
<p><img src="figure/ukb_bc_results.Rmd/results_75K_accuracy-1.png" width="1248" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-results_75K_accuracy-1">
Past versions of results_75K_accuracy-1.png
</button>
</p>
<div id="fig-results_75K_accuracy-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/morgantelab/mr_mash_rss_analysis/blob/60889dd871514919b74d43907d52d9ee47e06821/docs/figure/ukb_bc_results.Rmd/results_75K_accuracy-1.png" target="_blank">60889dd</a>
</td>
<td>
fmorgante
</td>
<td>
2024-02-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The percentage change in <span class="math inline">\(R^2\)</span> of
<em>mr.mash.rss init prior</em> over <em>BayesR</em> by trait is:</p>
<pre class="r"><code>perc_change_bayesR_fm_prior</code></pre>
<pre><code>        WBC_count         RBC_count       Haemoglobin               MCV 
     3.2036534199      6.3578037992     -0.0001688431      1.3798038543 
              RDW    Platelet_count      Plateletcrit               PDW 
     5.7435803143      3.9794938660      2.8926057502      1.5244281537 
  Lymphocyte_perc     Monocyte_perc  Neutrophill_perc  Eosinophill_perc 
     9.1073094157     -3.0325083273     11.5823290872     -2.4125973315 
   Basophill_perc Reticulocyte_perc              MSCV          HLR_perc 
    10.2058431304      7.0892935140      7.3419703787      3.0189283421 </code></pre>
<p>The overall percentage change in <span
class="math inline">\(R^2\)</span> of <em>mr.mash.rss init prior</em>
over <em>BayesR</em> across traits is 4.2488605.</p>
<!-- The results show that *mr.mash.rss init* improves accuracy over *LDpred2* more substantially with a smaller -->
<!-- sample size, highlighting the scenario where a multivariate analysis is really advantageous. -->
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.2 (2021-11-01)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Rocky Linux 8.5 (Green Obsidian)

Matrix products: default
BLAS/LAPACK: /opt/ohpc/pub/libs/gnu9/openblas/0.3.7/lib/libopenblasp-r0.3.7.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] cowplot_1.1.1 ggplot2_3.4.3

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.11       highr_0.10        pillar_1.9.0      compiler_4.1.2   
 [5] bslib_0.5.0       later_1.3.1       jquerylib_0.1.4   git2r_0.32.0     
 [9] workflowr_1.7.0   tools_4.1.2       digest_0.6.33     gtable_0.3.3     
[13] jsonlite_1.8.7    evaluate_0.21     lifecycle_1.0.3   tibble_3.2.1     
[17] pkgconfig_2.0.3   rlang_1.1.1       cli_3.6.1         rstudioapi_0.15.0
[21] yaml_2.3.7        xfun_0.39         fastmap_1.1.1     withr_2.5.0      
[25] dplyr_1.1.2       stringr_1.5.0     knitr_1.43        generics_0.1.3   
[29] fs_1.6.3          vctrs_0.6.3       sass_0.4.7        tidyselect_1.2.0 
[33] rprojroot_2.0.3   grid_4.1.2        glue_1.6.2        R6_2.5.1         
[37] fansi_1.0.4       rmarkdown_2.23    farver_2.1.1      magrittr_2.0.3   
[41] whisker_0.4.1     scales_1.2.1      promises_1.2.0.1  htmltools_0.5.5  
[45] colorspace_2.1-0  httpuv_1.6.11     labeling_0.4.2    utf8_1.2.3       
[49] stringi_1.7.12    munsell_0.5.0     cachem_1.0.8     </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
