---
title: "Analysis of blood cell traits in UK Biobank"
author: "Fabio Morgante & Deborah Kunkel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r set options}
###Load libraries
library(ggplot2)
library(cowplot)

prefix <- "output/prediction_accuracy/ukb_bc"
```

## Introduction

The goal of this analysis is to benchmark the newly developed *mr.mash.rss* (aka *mr.mash* with summary 
data) against already existing methods in the task of predicting phenotypes from genotypes using 
only summary data. Here we describe the analysis of 16 blood cell traits in UK Biobank. 

## Description of the analysis

Two different methods were fitted:

* [LDpred2](https://academic.oup.com/bioinformatics/article/36/22-23/5424/6039173) per-chromosome with 
the auto option, 1000 iterations (after 500 burn-in iterations), $h^2$ initialized using an estimate from 
LD Score regression (LDSC) and $p$ initialized using the same grid as in the original paper. 
NB this is a univariate method.

* _mr.mash.rss_ per-chromosome, with both canonical and data-driven covariance matrices computed 
as described in the [mvSuSiE paper](https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1),
updating the (full rank) residual covariance and the mixture weights, without standardizing the variables. 
The residual covariance was initialized as in the [mvSuSiE paper](https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1)
and the mixture weights were initialized as 90\% of the weight on the null component and 10\% of the
weight split equally across the remaining components. The phenotypic covariance was computed
as the sample covariance using the individual-level data. NB this is a multivariate method.

Prediction accuracy was evaluated as the $R^2$ of the regression of true phenotypes on the predicted 
phenotypes. This metric as the attractive property that its upper bound is $h_g^2$.

## Results

```{results_accuracy, fig.height=10, fig.width=13}
dat_ldsc <- matrix(as.numeric(NA), 5, 16)

for(i in 1:5){
  for(s in 1:16){
    dat_ldsc[i, s] <- readRDS(paste0("Desktop/ukb_bc_chrAll_ldsc_fit_trait", s, "_", i, ".rds"))["h2"]
  }
  
  if(i>1){
    dat_mrmash <- rbind(dat_mrmash, rep(as.numeric(NA), 16))
    #dat_mrmash <- rbind(dat_mrmash, readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_pred_acc_", i, ".rds"))$r2)
    dat_ldpred2 <- rbind(dat_ldpred2, readRDS(paste0(prefix, "_ldpred2_auto_pred_acc_", i, ".rds"))$r2)
  } else {
    dat_mrmash <- rep(as.numeric(NA), 16)
    #dat_mrmash <- readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_pred_acc_", i, ".rds"))$r2
    dat_ldpred2 <- readRDS(paste0(prefix, "_ldpred2_auto_pred_acc_", i, ".rds"))$r2
  }
}

linez <- data.frame(trait=factor(c("1", "2", "3", "4", "5", "6", "7", "8",
                            "9", "10", "11", "12", "13", "14", "15", "16"),
                            levels = c("1", "2", "3", "4", "5", "6", "7", "8",
                                       "9", "10", "11", "12", "13", "14", "15", "16")),
                    h2=colMeans(dat_ldsc))

r2 <- c(as.vector(dat_mrmash), as.vector(dat_ldpred2))
method <- rep(c("mr_mash_rss", "ldpred2_auto"), each=80)
trait <- rep(c("1", "2", "3", "4", "5", "6", "7", "8",
               "9", "10", "11", "12", "13", "14", "15", "16",
               "1", "2", "3", "4", "5", "6", "7", "8",
               "9", "10", "11", "12", "13", "14", "15", "16"), each=5)

res <- data.frame(method, trait, r2)

res <- res[which(res$method %in% c("ldpred2_auto", "mr_mash_rss")), ]

res <- transform(res,
                 method=factor(method, levels=c("ldpred2_auto", "mr_mash_rss"),
                               labels=c("LDpred2", "mr_mash_rss")),
                 trait=factor(trait, levels=c("1", "2", "3", "4", "5", "6", "7", "8",
                                              "9", "10", "11", "12", "13", "14", "15", "16")))

p_methods_shared <- ggplot(res, aes(x = trait, y = r2, fill = method)) +
  geom_boxplot(color = "black", outlier.size = 1, width = 0.85) +
  stat_summary(fun=mean, geom="point", shape=23,
               position = position_dodge2(width = 0.87,   
                                          preserve = "single")) +
  scale_fill_manual(values = c("pink", "red", "green", "blue")) +
  labs(x = "Trait", y = expression(italic(R)^2), fill="Method", title="") + 
  facet_grid(~trait, scales="free_x") +
  geom_hline(aes(yintercept = h2), linez) +
  theme_cowplot(font_size = 18)

print(p_methods_shared)

```
