---
title: "Analysis of blood cell traits in UK Biobank"
author: "Fabio Morgante & Deborah Kunkel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

```{r set options}
###Load libraries
library(ggplot2)
library(cowplot)

prefix <- "output/prediction_accuracy/ukb_bc"
prefix_2 <- "output/prediction_accuracy/ukb_bc_sampled"
```

## Introduction

The goal of this analysis is to benchmark the newly developed *mr.mash.rss* (aka *mr.mash* with summary 
data) against already existing methods in the task of predicting phenotypes from genotypes using 
only summary data. After performing several simulations, we analyzed analysis 16 blood cell traits in UK Biobank. 

## Description of the analysis

Following the [mvSuSiE paper](https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1), we selected 16 blood
cell traits from the total available in the UK Biobank haematology data collection and after the filtering
steps, we were left with 244,049 samples.

A 5-fold cross validation scheme was employed, whereby the samples were randomly divided in 5 disjoint sets.
At each iteration, 4 sets were used as training set and 1 set was used as test set.

Summary statistics were computed in the training set by first regressing out the effect of sex, age at recruitment,
age $\times$ age, assessment centre, and genotype measurement batch, and the top 10 genotype PCs using a linear
model. Then, we ran a GWAS using a simple linear regression on the quantile normalized residuals from the previous
step.

The sparse LD matrix was computed in the training set as described in the [LDpred2 paper](https://academic.oup.com/bioinformatics/article/36/22-23/5424/6039173)

For the majority of the analyses, we used a use a set of 1,054,330 HapMap3 variants. Only for the computation
of the data-driven covariance matrices for *mr.mash.rss* we used a larger sets of variants as described in the
[mvSuSiE paper](https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1).

Two different methods were fitted to the summary statistics:

* [LDpred2](https://academic.oup.com/bioinformatics/article/36/22-23/5424/6039173) per-chromosome with
the auto option, 1000 iterations (after 500 burn-in iterations), $h^2$ initialized using an estimate from
LD Score regression (LDSC) and $p$ initialized using the same grid as in the original paper.
NB this is a univariate method.
* [BayesR](https://www.nature.com/articles/s41467-019-12653-0) per-chromosome, with 5000 iterations,
1000 burn-in iterations, thinning factor of 5, $\pi$ (i.e., the proportion of causal variants)
initialized as 0.0001, $h^2$ initialized as 0.1, and other default parameters. NB this is a univariate method.
We used the implementation in the *qgg* R package.
* _mr.mash.rss_ per-chromosome, with both canonical and data-driven covariance matrices computed
similar to the [mvSuSiE paper](https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1). However,
variants were finemapped (via SuSiE-RSS) to select strong and random sets for the data-driven matrices. 
The residual covariance was estimated as in the mvSuSiE paper and fixed. The mixture weights were 
initialized as 90\% of the weight on the null component and 10\% of the weight split equally across 
the remaining components and then updated in our algorithm. The phenotypic covariance was computed as 
the sample covariance using the individual-level data. The posterior means of regression coefficients 
were initialized using the estimates from either *LDpred2* or *BayesR*). The genetic variants 
were not standardized. NB this is a multivariate method.

Prediction accuracy was evaluated as the $R^2$ of the regression of (quantile normalized) true phenotypes
on the predicted phenotypes in the test set. This metric as the attractive property that its upper bound
is $h_g^2$. Here we report the results for each fold.

## Results with 244K individuals

```{r results_244K_accuracy, fig.height=10, fig.width=13}
dat_ldsc <- matrix(as.numeric(NA), 5, 16)

for(i in 1:5){
  for(s in 1:16){
    dat_ldsc[i, s] <- readRDS(paste0("output/ldsc_fit/ukb_bc_chrAll_ldsc_fit_trait", s, "_", i, ".rds"))["h2"]
  }

  if(i>1){
    dat_mrmash <- rbind(dat_mrmash, readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_V_all_chr_pred_acc_", i, ".rds"))$r2)
    dat_mrmash_init <- rbind(dat_mrmash_init, readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_V_all_chr_init_pred_acc_", i, ".rds"))$r2)
    dat_mrmash_init_fm_prior <- rbind(dat_mrmash_init_fm_prior, readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_V_all_chr_init_prior_finemapped_pred_acc_", i, ".rds"))$r2)
    dat_mrmash_bayesR_init_fm_prior <- rbind(dat_mrmash_bayesR_init_fm_prior, readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_V_all_chr_bayesR_init_prior_finemapped_pred_acc_", i, ".rds"))$r2)
    dat_ldpred2 <- rbind(dat_ldpred2, readRDS(paste0(prefix, "_ldpred2_auto_pred_acc_", i, ".rds"))$r2)
    dat_bayesR <- rbind(dat_bayesR, readRDS(paste0(prefix, "_bayesR_pred_acc_", i, ".rds"))$r2)
  } else {
    dat_mrmash <- readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_V_all_chr_pred_acc_", i, ".rds"))$r2
    dat_mrmash_init <- readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_V_all_chr_init_pred_acc_", i, ".rds"))$r2
    dat_mrmash_init_fm_prior <- readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_V_all_chr_init_prior_finemapped_pred_acc_", i, ".rds"))$r2
    dat_mrmash_bayesR_init_fm_prior <- readRDS(paste0(prefix, "_mr_mash_rss_sparse_LD_V_all_chr_bayesR_init_prior_finemapped_pred_acc_", i, ".rds"))$r2
    dat_ldpred2 <- readRDS(paste0(prefix, "_ldpred2_auto_pred_acc_", i, ".rds"))$r2
    dat_bayesR <- readRDS(paste0(prefix, "_bayesR_pred_acc_", i, ".rds"))$r2
  }
}

means_mrmash <- colMeans(dat_mrmash)
means_mrmash_init <- colMeans(dat_mrmash_init)
means_mrmash_init_fm_prior <- colMeans(dat_mrmash_init_fm_prior)
means_mrmash_bayesR_init_fm_prior <- colMeans(dat_mrmash_bayesR_init_fm_prior)
means_ldpred2 <- colMeans(dat_ldpred2)
means_bayesR <- colMeans(dat_bayesR)

perc_change <- ((means_mrmash_init-means_ldpred2)/means_ldpred2)*100
perc_change_fm_prior <- ((means_mrmash_init_fm_prior-means_ldpred2)/means_ldpred2)*100
perc_change_bayesR_fm_prior <- ((means_mrmash_init_fm_prior-means_bayesR)/means_bayesR)*100
perc_change_bayesR_bayesR_init_fm_prior <- ((means_mrmash_bayesR_init_fm_prior-means_bayesR)/means_bayesR)*100

pheno <- readRDS("data/phenotypes/ukb_cleaned_bc_adjusted_pheno_test_1.rds")

names(perc_change) <- colnames(pheno)
names(perc_change_fm_prior) <- colnames(pheno)
names(perc_change_bayesR_fm_prior) <- colnames(pheno)
names(perc_change_bayesR_bayesR_init_fm_prior) <- colnames(pheno)

linez <- data.frame(trait=colnames(pheno),
                    h2=colMeans(dat_ldsc))

r2 <- c(as.vector(dat_mrmash), as.vector(dat_mrmash_init), as.vector(dat_mrmash_init_fm_prior),
        as.vector(dat_mrmash_bayesR_init_fm_prior), as.vector(dat_ldpred2), as.vector(dat_bayesR))
method <- rep(c("mr_mash_rss", "mr_mash_rss_init", "mr_mash_rss_init_fm_prior",
                "mr_mash_rss_bayesR_init_fm_prior", "ldpred2_auto", "bayesR"), each=80)
trait <- rep(colnames(pheno), each=5)

res <- data.frame(method, trait, r2)

#res <- res[which(res$method %in% c("bayesR", "ldpred2_auto", "mr_mash_rss", "mr_mash_rss_init", "mr_mash_rss_init_fm_prior")), ]
res <- res[which(res$method %in% c("bayesR", "ldpred2_auto",
                                   "mr_mash_rss_bayesR_init_fm_prior")), ]

res <- transform(res,
                 method=factor(method, levels=c("bayesR", "ldpred2_auto", "mr_mash_rss_init_fm_prior",
                                                "mr_mash_rss_bayesR_init_fm_prior"),
                               labels=c("BayesR", "LDpred2", "mr.mash.rss init prior", "mr.mash.rss BayesR init prior")),
                 trait=factor(trait))

p_methods_shared <- ggplot(res, aes(x = trait, y = r2, fill = method)) +
  geom_boxplot(color = "black", outlier.size = 1, width = 0.85) +
  stat_summary(fun=mean, geom="point", shape=23,
               position = position_dodge2(width = 0.87,
                                          preserve = "single")) +
  scale_fill_manual(values = c("pink", "red", "blue", "green", "orange")) +
  labs(x = "Trait", y = expression(italic(R)^2), fill="Method", title="") +
  # facet_grid(~trait, scales="free_x") +
  # geom_hline(aes(yintercept = h2), linez) +
  theme_cowplot(font_size = 18) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

print(p_methods_shared)
```

The percentage change in $R^2$ of *mr.mash.rss init* over *LDpred2* by trait is:

```{r perc_change}
perc_change
```

The percentage change in $R^2$ of *mr.mash.rss BayesR init prior* over *BayesR* (the best univariate method) by trait is:

```{r perc_change_fm_prior}
perc_change_bayesR_bayesR_init_fm_prior
```

The overall percentage change in $R^2$ of *mr.mash.rss BayesR init prior* over *BayesR* across traits is `r mean(perc_change_bayesR_bayesR_init_fm_prior)`.

<!-- The results show that *mr.mash.rss init* improves accuracy over *LDpred2* for all but one trait. However, -->
<!-- in that trait the performance is very similar. In general, the improvement in performance is in line with -->
<!-- other papers such as the [PRS-CS paper](https://www.nature.com/articles/s41467-019-09718-5). -->


## Results with 75K individuals

We hypothesized that the advantage of a multivariate analysis would be more pronounced with a smaller sample
size. Thus, we randomly sampled 75,000 individuals from the 244,000 and repeated the analyses above.

```{r results_75K_accuracy, fig.height=10, fig.width=13}
for(i in 1:5){
  if(i>1){
    dat_mrmash_init_fm_prior <- rbind(dat_mrmash_init_fm_prior, readRDS(paste0(prefix_2, "_mr_mash_rss_sparse_LD_V_all_chr_init_prior_finemapped_pred_acc_", i, ".rds"))$r2)
    dat_mrmash_bayesR_init_fm_prior <- rbind(dat_mrmash_bayesR_init_fm_prior, readRDS(paste0(prefix_2, "_mr_mash_rss_sparse_LD_V_all_chr_bayesR_init_prior_finemapped_pred_acc_", i, ".rds"))$r2)
    dat_ldpred2 <- rbind(dat_ldpred2, readRDS(paste0(prefix_2, "_ldpred2_auto_pred_acc_", i, ".rds"))$r2)
    dat_bayesR <- rbind(dat_bayesR, readRDS(paste0(prefix_2, "_bayesR_pred_acc_", i, ".rds"))$r2)
  } else {
    dat_mrmash_init_fm_prior <- readRDS(paste0(prefix_2, "_mr_mash_rss_sparse_LD_V_all_chr_init_prior_finemapped_pred_acc_", i, ".rds"))$r2
    dat_mrmash_bayesR_init_fm_prior <- readRDS(paste0(prefix_2, "_mr_mash_rss_sparse_LD_V_all_chr_bayesR_init_prior_finemapped_pred_acc_", i, ".rds"))$r2
    dat_ldpred2 <- readRDS(paste0(prefix_2, "_ldpred2_auto_pred_acc_", i, ".rds"))$r2
    dat_bayesR <- readRDS(paste0(prefix_2, "_bayesR_pred_acc_", i, ".rds"))$r2
  }
}

means_mrmash_init_fm_prior <- colMeans(dat_mrmash_init_fm_prior)
means_mrmash_bayesR_init_fm_prior <- colMeans(dat_mrmash_bayesR_init_fm_prior)
means_ldpred2 <- colMeans(dat_ldpred2)
means_bayesR <- colMeans(dat_bayesR)

perc_change_fm_prior <- ((means_mrmash_init_fm_prior-means_ldpred2)/means_ldpred2)*100
perc_change_bayesR_fm_prior <- ((means_mrmash_init_fm_prior-means_bayesR)/means_bayesR)*100
perc_change_bayesR_bayesR_init_fm_prior <- ((means_mrmash_bayesR_init_fm_prior-means_bayesR)/means_bayesR)*100

pheno <- readRDS("data/phenotypes/ukb_cleaned_bc_sampled_adjusted_pheno_test_1.rds")

names(perc_change_fm_prior) <- colnames(pheno)
names(perc_change_bayesR_fm_prior) <- colnames(pheno)
names(perc_change_bayesR_bayesR_init_fm_prior) <- colnames(pheno)


r2 <- c(as.vector(dat_mrmash_init_fm_prior), as.vector(dat_mrmash_bayesR_init_fm_prior), as.vector(dat_ldpred2), as.vector(dat_bayesR))
method <- rep(c("mr_mash_rss_init_fm_prior", "mr_mash_rss_bayesR_init_fm_prior", "ldpred2_auto", "bayesR"), each=80)
trait <- rep(colnames(pheno), each=5)

res <- data.frame(method, trait, r2)

res <- res[which(res$method %in% c("bayesR", "ldpred2_auto", "mr_mash_rss_init_fm_prior", "mr_mash_rss_bayesR_init_fm_prior")), ]

res <- transform(res,
                 method=factor(method, levels=c("bayesR", "ldpred2_auto", "mr_mash_rss_init_fm_prior", "mr_mash_rss_bayesR_init_fm_prior"),
                               labels=c("BayesR", "LDpred2", "mr.mash.rss init prior", "mr.mash.rss BayesR init prior")),
                 trait=factor(trait))

p_methods_shared <- ggplot(res, aes(x = trait, y = r2, fill = method)) +
  geom_boxplot(color = "black", outlier.size = 1, width = 0.85) +
  stat_summary(fun=mean, geom="point", shape=23,
               position = position_dodge2(width = 0.87,
                                          preserve = "single")) +
  scale_fill_manual(values = c("pink", "red", "blue", "green")) +
  labs(x = "Trait", y = expression(italic(R)^2), fill="Method", title="") +
  # facet_grid(~trait, scales="free_x") +
  # geom_hline(aes(yintercept = h2), linez) +
  theme_cowplot(font_size = 18) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

print(p_methods_shared)
```

The percentage change in $R^2$ of *mr.mash.rss BayesR init prior* over *BayesR* by trait is:

```{r perc_change_fm_prior_75k}
perc_change_bayesR_bayesR_init_fm_prior
```


The overall percentage change in $R^2$ of *mr.mash.rss BayesR init prior* over *BayesR* across traits is `r mean(perc_change_bayesR_bayesR_init_fm_prior)`.

The results show that *mr.mash.rss BayesR init* improves accuracy over *BayesR* more substantially with a smaller
sample size, highlighting the scenario where a multivariate analysis is really advantageous.
