---
title: "Simulations with UK Biobank genotypes and different effect structure across traits"
author: "Fabio Morgante & Deborah Kunkel"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  workflowr::wflow_html:
    toc: true
editor_options:
  chunk_output_type: console
---

## Introduction

The goal of this analysis is to benchmark the newly developed *mr.mash.rss* (aka *mr.mash* with summary 
statistics) against already existing methods in the task of predicting phenotypes from genotypes using 
only summary data. To do so, we used real genotypes from the array data of the UK Biobank. We randomly 
sampled 105,000 nominally unrelated ($r_A$ < 0.025 between any pair) individuals of European ancestry 
(i.e., Caucasian and white British fields). After retaining variants with minor allele frequency 
(MAF) > 0.01, minor allele count (MAC) > 5 genotype missing rate < 0.1 and Hardy-Weinberg Equilibrium (HWE) 
test p-value > $1 *10^{-10}$, our data consisted of 595,071 genetic variants (i.e., our predictors). 
Missing genotypes were imputed with the mean genotype for the respective genetic variant.

The linkage disequilibrium (LD) matrices were computed using 146,288 nominally unrelated ($r_A$ < 0.025 
between any pair) individuals of European ancestry (i.e., Caucasian and white British fields), that did 
not overlap with the 105,000 individuals used for the rest of the analyses.

For each simulation replicate, we randomly sampled 5,000 (out of the 105,000) individuals to be the test 
set. The test set was only used to evaluate prediction accuracy. All the other steps were carried out on 
the training set of 100,000 individuals.

We simulated 5 traits (i.e., our responses) by randomly sampling 5,000 variants (out of the total of 
595,071) to be causal, with different effect sharing structures across traits (see below). The genetic
effects explain 50\% of the total per-trait variance. The residuals are uncorrelated across traits.

Summary statistics (i.e., effect size and its standard error) were obtained by univariate simple linear 
regression of each trait on each variant, one at a time. Traits and variants were not standardized.

Three different methods were fitted:

* [LDpred2](https://academic.oup.com/bioinformatics/article/36/22-23/5424/6039173) per-chromosome with 
the auto option, 500 iterations (after 500 burn-in iterations), $h^2$ initialized as 0.5/22 and $p$ 
initialized using the same grid as in the original paper.
* _LDpred2_ genome-wide with the auto option, 500 iterations (after 500 burn-in iterations), $h^2$ 
initialized as 0.5/22 (WRONG) and $p$ initialized using the same grid as in the original paper.
* _mr.mash.rss_ per-chromosome, with only data-driven covariance matrices computed as described in the
[mr.mash paper](https://www.biorxiv.org/content/10.1101/2022.11.22.517471v4.abstract), updating the (full 
rank) residual covariance and the mixture weights, without standardizing the variables.



## Equal effects scenario

```{r, fig.height=12, fig.width=15}
###Load libraries
library(ggplot2)
library(cowplot)

repz <- 1:20
prefix <- "output/prediction_accuracy/ukb_caucasian_white_british_unrel_100000"
scenarioz <- "equal_effects_indep_resid"
methodz <- c("mr_mash_rss", "ldpred2_auto", "ldpred2_auto_gwide")
metric <- "r2"
traitz <- 1:5

i <- 0

n_col <- 6
n_row <- length(repz) * length(scenarioz) * length(methodz) * length(traitz)
res <- as.data.frame(matrix(NA, ncol=n_col, nrow=n_row))
colnames(res) <- c("rep", "scenario", "method", "trait", "metric", "score")



for(sce in scenarioz){
  for(met in methodz){
    for(repp in repz){
      dat <- readRDS(paste0(prefix, "_", sce, "_", met, "_pred_acc_", repp, ".rds"))
      
      for(trait in traitz){
        i <- i + 1
        
        res[i, 1] <- repp
        res[i, 2] <- sce
        res[i, 3] <- met
        res[i, 4] <- trait
        res[i, 5] <- metric
        res[i, 6] <- dat$r2[trait]
      }
    }
  }
}

res <- transform(res, scenario=as.factor(scenario),
                      method=as.factor(method),
                      trait=as.factor(trait))

p_methods_shared <- ggplot(res, aes(x = trait, y = score, fill = method)) +
  geom_boxplot(color = "black", outlier.size = 1, width = 0.85) +
  stat_summary(fun=mean, geom="point", shape=23,
               position = position_dodge2(width = 0.87,   
                                          preserve = "single")) +
  ylim(0.3, 0.51) +
  scale_fill_manual(values = c("pink", "red", "green"))+ #, labels = c("g-lasso", "smt-lasso", "e-net")) +
  labs(x = "Trait", y = expression(italic(R)^2), title = "Equal effects", fill="Method") +
  geom_hline(yintercept=0.5, linetype="dotted", linewidth=1, color = "black") +
  theme_cowplot(font_size = 16) +
  theme(plot.title = element_text(hjust = 0.5, size=14))

print(p_methods_shared)
```

